dl/src/main/scala/com/intel/analytics/bigdl/nn/TemporalConvolution.scala

```diff
@@ -0,0 +1,381 @@
+  val gradWeight: Tensor[T] = Tensor[T](outputFrameSize, inputFrameSize * kernelW)
+  val gradBias: Tensor[T] = Tensor[T](outputFrameSize)
+
+  override def reset(): Unit = {
+    initMethod match {
+      case Default =>
+        val stdv = 1.0 / math.sqrt(kernelW * inputFrameSize)
+        weight.apply1(_ => ev.fromType[Double](RNG.uniform(0, 1) * 2 * stdv - stdv))
+        bias.apply1(_ => ev.fromType[Double](RNG.uniform(0, 1) * 2 * stdv - stdv))
+      case Xavier =>
+        val fanIn = inputFrameSize * kernelW
+        val fanOut = outputFrameSize * kernelW
+        val stdv = math.sqrt(6.0 / (fanIn + fanOut))
+        weight.apply1(_ => ev.fromType[Double](RNG.uniform(-stdv, stdv)))
+        bias.fill(ev.fromType(0))
+      case _ => throw new IllegalArgumentException()
+    }
+    zeroGradParameters()
+  }
+
+  override def updateOutput(input: Tensor[T]): Tensor[T] = {
+    require(input.dim() == 2 || input.dim() == 3,
+      "TemporalConvolution: 2D or 3D(batch mode) tensor expected for input, " +
+        s"but got ${input.dim()}")
+    require(input.isContiguous())
+
+    val dimSeq = if (input.dim() == 2) 1 else 2
+    val dimFeat = if (input.dim() == 2) 2 else 3
+    val nInputFrame = input.size(dimSeq)
+    var nOutputFrame = (nInputFrame - kernelW) / strideW + 1
+    val inputWindow = Tensor[T]()
+    var outputWindow = Tensor[T]()
+
+    require(input.size(dimFeat) == inputFrameSize, "Invalid input frame size. Got: " +
+      s"${input.size(dimFeat)}, Expected: $inputFrameSize")
+    require(nOutputFrame >= 1, "Input sequence smaller than kernel size. Got: " +
+      s"$nInputFrame, Expected: $kernelW")
+
+    if (input.dim() == 2) {
+      // Time x Feature
+      output.resize(Array(nOutputFrame, outputFrameSize))
+      var j = 1
+      while (j <= nOutputFrame) {
+        outputWindow = output.select(1, j)
+        outputWindow.copy(bias)
+        j += 1
+      }
+
+      j = 0
+      while (nOutputFrame > 0) {
+        val outputFrameStride = (kernelW - 1) / strideW + 1
+        val inputFrameStride = outputFrameStride * strideW
+        val nFrame = (nInputFrame - j * strideW - kernelW) / inputFrameStride + 1
+        nOutputFrame -= nFrame
+
+        inputWindow.set(input.storage(), input.storageOffset() + j * strideW * input.size(2),
+          Array(nFrame, kernelW * input.size(2)), Array(inputFrameStride * input.size(2), 1))
+        outputWindow.set(output.storage(), output.storageOffset() + j * output.size(2),
+          Array(nFrame, output.size(2)), Array(outputFrameStride * output.size(2), 1))
+
+        val weightT = weight.transpose(1, 2)
+        outputWindow.addmm(ev.fromType[Int](1), outputWindow,
+          ev.fromType[Int](1), inputWindow, weightT)
+        j += 1
+      }
+
+    } else {
+      // Batch x Time x Feature
+      val batchSize = input.size(1)
+      output.resize(Array(batchSize, nOutputFrame, outputFrameSize))
+
+      var i = 0
+      while (i < batchSize) {
```

Comment:
If samples in different batch are not dependent on one another, parallel these calculations? Please refer to SpatialConvolution.scala updateOutput